File: iosApp/iosApp/Analytics/FirebaseAnalyticsService.swift
//
//  FirebaseAnalyticsService.swift
//  DevFest Nantes
//
//  Created by Stéphane Rihet on 28/09/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import Firebase
import shared

class FirebaseAnalyticsService: AnalyticsService {
    ///Singleton FirebaseAnalytics
    static let shared = FirebaseAnalyticsService()
    
    required internal init() {}
    
    // MARK: - Events
    
    ///Method called when a session is added or removed from favorites.
    func eventBookmark(page: AnalyticsPage, sessionId: String, bookmarked: Bool) {
        if bookmarked {
            Analytics.logEvent(AnalyticsEvent.addToBookmarks.description(), parameters: [
                AnalyticsParam.fromPage.description(): page.description(),
                AnalyticsParam.sessionId.description(): sessionId
                ]
            )
        } else {
            Analytics.logEvent(AnalyticsEvent.removeFromBookmarks.description(), parameters: [
                AnalyticsParam.fromPage.description(): page.description(),
                AnalyticsParam.sessionId.description(): sessionId
                ]
            )
        }
    }
    
    ///Method called when a filter is selected (To Do).
    func eventFilter() {
        //TO DO
    }
    
    ///Method called when clicking on the code of conduct button.
    func eventLinkCodeOfConductOpened() {
        Analytics.logEvent(AnalyticsEvent.linkCodeOfConductOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the DevFest website button.
    func eventLinkDevFestWebsiteOpened() {
        Analytics.logEvent(AnalyticsEvent.linkDevfestWebsiteOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the facebook button.
    func eventLinkFacebookOpened() {
        Analytics.logEvent(AnalyticsEvent.linkFacebookOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the twitter button.
    func eventLinkTwitterOpened() {
        Analytics.logEvent(AnalyticsEvent.linkTwitterOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the github button.
    func eventLinkGithubOpened() {
        Analytics.logEvent(AnalyticsEvent.linkGithubOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the linkedin button.
    func eventLinkLinkedinOpened() {
        Analytics.logEvent(AnalyticsEvent.linkLinkedinOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the local communities button.
    func eventLinkLocalCommunitiesOpened() {
        Analytics.logEvent(AnalyticsEvent.linkLocalCommunitiesOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the partner's logo.
    func eventLinkPartnerOpened(partnerName: String) {
        Analytics.logEvent(AnalyticsEvent.linkPartnerOpened.description(), parameters: [
            AnalyticsParam.partnerName.description(): partnerName
            ]
        )
    }
    
    ///Method called when clicking on the support button.
    func eventLinkSupportOpened() {
        Analytics.logEvent(AnalyticsEvent.linkSupportOpened.description(), parameters: [:] )
    }
    
    ///Method called when clicking on the youtube button.
    func eventLinkYoutubeOpened() {
        Analytics.logEvent(AnalyticsEvent.linkYoutubeOpened.description(), parameters: [:] )
    }

    ///Method called when clicking on the feedback button.
    func eventFeedbackClicked(openFeedbackId: String) {
        Analytics.logEvent(AnalyticsEvent.feedbackClicked.description(), parameters: [
            AnalyticsParam.openfeedbackId.description(): String(openFeedbackId)
            ]
        )
    }
    
    ///Method called when selecting a session.
    func eventSessionOpened(sessionId: String) {
        Analytics.logEvent(AnalyticsEvent.linkPartnerOpened.description(), parameters: [
            AnalyticsParam.sessionId.description(): String(sessionId)
            ]
        )
    }

    func eventSpeakerOpened(speakerId: String) {
        Analytics.logEvent(AnalyticsEvent.speakerOpened.description(), parameters: [
            AnalyticsParam.speakerId.description(): String(speakerId)
            ]
        )
    }
    
    
    ///Method called when a speaker's social network is clicked.
    func eventSpeakerSocialLinkOpened(speakerId: String, type: SocialType) {
        Analytics.logEvent(AnalyticsEvent.speakerSocialLinkOpened.description(), parameters: [
            AnalyticsParam.speakerId.description(): speakerId,
            AnalyticsParam.socialType.description(): type
            ]
        )
    }
    
    ///Method called when the floorplan is clicked.
    func eventVenueFloorPlanClicked() {
        Analytics.logEvent(AnalyticsEvent.venueFloorPlanClicked.description(), parameters: [:])
    }
    
    ///Method called when clicking on the navigation button.
    func eventVenueNavigationClicked() {
        Analytics.logEvent(AnalyticsEvent.venueNavigationClicked.description(), parameters: [:])
    }
    
    // MARK: - Page
    ///Method called when initializing a page.
    func pageEvent(page: AnalyticsPage, className: String?) {
        Analytics.logEvent(AnalyticsEventScreenView,
                           parameters: [AnalyticsParameterScreenName: page.description(), AnalyticsParameterScreenClass: className!])
    }
}

File: androidApp/src/main/java/com/gdgnantes/devfest/androidapp/services/FirebaseAnalyticsService.kt
package com.gdgnantes.devfest.androidapp.services

import com.gdgnantes.devfest.analytics.AnalyticsEvent
import com.gdgnantes.devfest.analytics.AnalyticsPage
import com.gdgnantes.devfest.analytics.AnalyticsParam
import com.gdgnantes.devfest.analytics.AnalyticsService
import com.gdgnantes.devfest.model.SocialType
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.logEvent
import timber.log.Timber
import javax.inject.Inject

class FirebaseAnalyticsService @Inject constructor(private val firebaseAnalytics: FirebaseAnalytics) :
    AnalyticsService {
    override fun eventBookmark(page: AnalyticsPage, sessionId: String, bookmarked: Boolean) {
        if (bookmarked) {
            firebaseAnalytics.logEvent(AnalyticsEvent.ADD_TO_BOOKMARKS.toString()) {
                param(AnalyticsParam.SESSION_ID.toString(), sessionId)
                param(AnalyticsParam.FROM_PAGE.toString(), page.toString())
            }
        } else {
            firebaseAnalytics.logEvent(AnalyticsEvent.REMOVE_FROM_BOOKMARKS.toString()) {
                param(AnalyticsParam.SESSION_ID.toString(), sessionId)
                param(AnalyticsParam.FROM_PAGE.toString(), page.toString())
            }
        }
    }

    override fun eventFilter() {
        Timber.d("TODO: eventFilter")
    }

    override fun eventLinkCodeOfConductOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_CODE_OF_CONDUCT_OPENED.toString()) {}
    }

    override fun eventLinkDevFestWebsiteOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_DEVFEST_WEBSITE_OPENED.toString()) {}
    }

    override fun eventLinkFacebookOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_FACEBOOK_OPENED.toString()) {}
    }

    override fun eventLinkGithubOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_GITHUB_OPENED.toString()) {}
    }

    override fun eventLinkLinkedinOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_LINKEDIN_OPENED.toString()) {}
    }

    override fun eventLinkLocalCommunitiesOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_LOCAL_COMMUNITIES_OPENED.toString()) {}
    }

    override fun eventLinkPartnerOpened(partnerName: String) {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_PARTNER_OPENED.toString()) {
            param(AnalyticsParam.PARTNER_NAME.toString(), partnerName)
        }
    }

    override fun eventLinkSupportOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_SUPPORT_OPENED.toString()) {}
    }

    override fun eventLinkTwitterOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_TWITTER_OPENED.toString()) {}
    }

    override fun eventLinkYoutubeOpened() {
        firebaseAnalytics.logEvent(AnalyticsEvent.LINK_YOUTUBE_OPENED.toString()) {}
    }

    override fun eventNavigationClicked() {
        firebaseAnalytics.logEvent(AnalyticsEvent.NAVIGATION_CLICKED.toString()) {}
    }

    override fun eventSessionOpened(sessionId: String) {
        firebaseAnalytics.logEvent(AnalyticsEvent.SESSION_OPENED.toString()) {
            param(AnalyticsParam.SESSION_ID.toString(), sessionId)
        }
    }

    override fun eventSpeakerSocialLinkOpened(speakerId: String, type: SocialType) {
        firebaseAnalytics.logEvent(AnalyticsEvent.SPEAKER_SOCIAL_LINK_OPENED.toString()) {
            param(AnalyticsParam.SPEAKER_ID.toString(), speakerId)
            param(AnalyticsParam.SOCIAL_TYPE.toString(), type.name)
        }
    }

    override fun pageEvent(page: AnalyticsPage, className: String?) {
        firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, page.toString())
            className?.run { param(FirebaseAnalytics.Param.SCREEN_CLASS, className) }
        }
    }

}
